src/target/renesas.c

```diff
@@ -0,0 +1,938 @@
+#define RV40_FSTATR (RV40_BASE + 0x80U) /* Flash Status */
+
+#define RV40_FSTATR_DBFULL (1U << 10U) /* Data Buffer Full */
+#define RV40_FSTATR_RDY    (1U << 15U) /* Flash Ready */
+
+#define RV40_FSTATR_PRGERR    (1U << 12U) /* Programming Error */
+#define RV40_FSTATR_ERSERR    (1U << 13U) /* Erasure Error */
+#define RV40_FSTATR_ILGLERR   (1U << 14U) /* Illegal Command Error */
+#define RV40_FSTATR_OTERR     (1U << 20U) /* Other Error */
+#define RV40_FSTATR_SECERR    (1U << 21U) /* Security Error */
+#define RV40_FSTATR_FESETERR  (1U << 22U) /* FENTRY Setting Error */
+#define RV40_FSTATR_ILGCOMERR (1U << 23U) /* Illegal Command Error */
+
+#define RV40_FSADDR (RV40_BASE + 0x30U)
+
+#define RV40_FMEPROT        (RV40_BASE + 0x44U)
+#define RV40_FMEPROT_LOCK   (0xD901U)
+#define RV40_FMEPROT_UNLOCK (0xD900U)
+
+#define RV40_FENTRYR            (RV40_BASE + 0x84U)
+#define RV40_FENTRYR_KEY_OFFSET 8U
+#define RV40_FENTRYR_KEY        (0xAAU << RV40_FENTRYR_KEY_OFFSET)
+#define RV40_FENTRYR_PE_CF      (1U)
+#define RV40_FENTRYR_PE_DF      (1U << 7U)
+
+#define RV40_FCPSR         (RV40_BASE + 0xE0U)
+#define RV40_FCPSR_ESUSPMD 1U
+
+#define RENESAS_CF_END (0x00200000UL) /* End of Flash (maximum possible accross families) */
+
+static bool renesas_uid(target *t, int argc, const char **argv);
+
+const struct command_s renesas_cmd_list[] = {
+	{"uid", (cmd_handler)renesas_uid, "Prints unique number"},
+	{NULL, NULL, NULL},
+};
+
+struct renesas_priv_s {
+	uint8_t pnr[17]; /* 16-byte PNR + 1-byte null termination */
+	pnr_series_t series;
+	uint32_t flash_root_table; /* if applicable */
+};
+
+static uint32_t renesas_fmifrt_read(target *t)
+{
+	return target_mem_read32(t, RENESAS_FMIFRT);
+}
+
+static void renesas_uid_read(target *t, const uint32_t base, uint8_t *uid)
+{
+	uint32_t uidr[4];
+	for (size_t i = 0U; i < 4U; i++)
+		uidr[i] = target_mem_read32(t, base + i * 4U);
+
+	for (size_t i = 0U; i < 16U; i++)
+		uid[i] = uidr[i / 4U] >> (i % 4U) * 8U;
+}
+
+static bool renesas_pnr_read(target *t, const uint32_t base, uint8_t *pnr)
+{
+	uint32_t pnrr[4];
+	for (size_t i = 0U; i < 4U; i++)
+		pnrr[i] = target_mem_read32(t, base + i * 4U);
+
+	if (base == RENESAS_FIXED1_PNR) {
+		/* Renesas... look what you made me do...  */
+		/* see 'Part numbering scheme' note for context */
+		for (size_t i = 0U; i < 13U; i++)
+			pnr[i] = pnrr[3U - (i + 3U) / 4U] >> (24U - ((i + 3U) % 4U) * 8U);
+		memset(pnr + 13, 0x20, 3);
+	} else {
+		for (size_t i = 0; i < 16; i++)
+			pnr[i] = pnrr[i / 4U] >> (i % 4U) * 8U;
+	}
+
+	/* all Renesas mcus start with 'R7', sanity check */
+	return pnr[0] == 'R' && pnr[1] == '7';
+}
+
+static pnr_series_t renesas_series(const uint8_t *pnr)
+{
+	uint32_t series = 0;
+	for (size_t i = 0; i < 4; i++)
+		series = (series << 8) | pnr[PNR_FAMILY_INDEX + i];
+
+	return (pnr_series_t)series;
+}
+
+static uint32_t renesas_flash_size(const uint8_t *pnr)
+{
+	switch (pnr[PNR_MEMSIZE_INDEX]) {
+	case PNR_MEMSIZE_16KB:
+		return 16UL * 1024UL;
```

Comment:
Rather than suffixing the individual constant values, this looks like a good candidate for expressing as `UINT32_C(16 * 1024)`